Assets/MRTK/SDK/Features/Utilities/LRUCache.cs

```diff
@@ -0,0 +1,336 @@
+                leastRecentEntry.Next = null;
+            }
+        }
+
+        /// <summary>
+        /// Removes an entry from the cache by its key.
+        /// </summary>
+        /// <param name="key">The key for the cached item.</param>
+        /// <returns>True if the item exists, false if item does not exists in the cache.</returns>
+        /// <exception>
+        /// <see cref="ArgumentNullException"/> if <paramref name="key"/> is null.
+        /// </exception>
+        public bool Remove(TKey key)
+        {
+            if (keyToEntryTable.TryGetValue(key, out var cacheEntry))
+            {
+                keyToEntryTable.Remove(cacheEntry.Key);
+                RemoveEntryFromList(cacheEntry);
+
+                // update most recent entry
+                if (cacheEntry == mostRecentEntry)
+                {
+                    mostRecentEntry = cacheEntry.Next;
+                }
+
+                // update least recent entry
+                if (cacheEntry == leastRecentEntry)
+                {
+                    leastRecentEntry = cacheEntry.Previous;
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Clear all entry out of the cache.
+        /// </summary>
+        public void Clear()
+        {
+            // Clear all entry stored in the dictionary.
+            keyToEntryTable.Clear();
+
+            // Iterate through the list and unhook all the pointers in the list.
+            while (mostRecentEntry != null)
+            {
+                var nextEntry = mostRecentEntry.Next;
+                mostRecentEntry.Next = null;
+
+                if (nextEntry != null)
+                {
+                    nextEntry.Previous = null;
+                }
+
+                mostRecentEntry = nextEntry;
+            }
+
+            leastRecentEntry = null;
+        }
+
+        /// <summary>
+        /// Returns an enumerator that iterates through cached items in order of most to least recently used.
+        /// </summary>
+        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
+        {
+            if (keyToEntryTable.Count == 0)
+            {
+                yield break;
+            }
+
+            for (var entry = mostRecentEntry; entry != null; entry = entry.Next)
+            {
+                yield return new KeyValuePair<TKey, TValue>(entry.Key, entry.Value);
+            }
+        }
+
+        /// <summary>
+        /// Update the doubly linked list to bring the given entry to the front of the list, with the front being the
+        /// most recently access entry.
+        /// </summary>
+        /// <param name="entry">The entry to bring to the front of the list</param>
+        private void MakeCacheEntryRecent(CacheEntry entry)
+        {
+            // Only perform operation if the entry is not already the most recently accessed item
+            if (entry != mostRecentEntry)
```

Comment:
Out of curiosity, do you happen to have any perf traces of all this cache management vs the current method of always calling `GetComponent`? I'm assuming this has been measured as less expensive, but it'd be interesting to see!

Reply:
I'll check with @CharlesWanMS, who made this change downstream

Reply:
Attaching the pre-optimized trace here:
![caching optimization pre](https://user-images.githubusercontent.com/47646014/151267168-66fe25ae-d775-4d00-bb75-60eb0054f587.PNG)

On it's own it may not seem much, but in our app we had spikes where those query are performed up to 12 times per frame for multiple frame. Since our app is CPU intensive, we had to hunt for saving where ever we could. I don't have a snap for post as all we did was running on device and didn't notice to spike while using the app.